scipy/optimize/minpack.py

```diff
@@ -449,12 +449,24 @@ def leastsq(func, x0, args=(), Dfun=None, full_output=0,
         return (retval[0], info)
 
 
-def _general_function(params, xdata, ydata, function):
-    return function(xdata, *params) - ydata
+def _wrap_func(func, xdata, ydata, weights):
+    if weights is None:
+        def func_wrapped(params):
+            return func(xdata, *params) - ydata
+    else:
+        def func_wrapped(params):
+            return weights * (func(xdata, *params) - ydata)
+    return func_wrapped
 
 
-def _weighted_general_function(params, xdata, ydata, function, weights):
-    return weights * (function(xdata, *params) - ydata)
+def _wrap_jac(jac, xdata, weights):
+    if weights is None:
+        def jac_wrapped(params):
+            return jac(xdata, *params)
+    else:
+        def jac_wrapped(params):
+            return weights[:, np.newaxis] * jac(xdata, *params)
```

Comment:
May need some special casing for sparse jacobians/np.matrix, for which `*` is matmul.